src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,399 @@
+            {
+                this.semanticModel = semanticModel;
+                this.baseType = baseType;
+            }
+        }
+
+
+        protected SyntaxNode? TryGetTargetNode(SyntaxNode root, TextSpan span)
+        {
+            var ancestors = root.FindToken(span.Start).GetAncestors<SyntaxNode>();
+
+            var node = ancestors.FirstOrDefault(n => n.Span.Contains(span) && n != root);
+            return node;
+        }
+
+        protected bool TryGetNode(SyntaxNode root, TextSpan span, SyntaxKind kind, SyntaxNode target, out SyntaxNode? node)
+        {
+            var ancestors = root.FindToken(span.Start).GetAncestors<SyntaxNode>();
+
+            node = ancestors.FirstOrDefault(n => n.Span.Contains(span) && n != root && n != target && n.IsKind(kind));
+            return node != null;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            if (root == null)
+            {
+                return;
+            }
+            var targetNodes = diagnostics.SelectAsArray(
+                d => TryGetTargetNode(root, d.Location.SourceSpan) is ExpressionSyntax node ? node : null);
+
+            await editor.ApplyExpressionLevelSemanticEditsAsync(
+                document, targetNodes,
+                (semanticModel, targetNode) => true,
+                (semanticModel, currentRoot, targetNode) =>
+                {
+                    if (GetTypeInfo(semanticModel, currentRoot, targetNode, cancellationToken, out var nodeType, out var conversionType, out var potentialConvTypes) &&
+                    nodeType != null && conversionType != null && !nodeType.Equals(conversionType) &&
+                        targetNode is ExpressionSyntax expression)
+                    {
+                        var castExpression = expression.Cast(conversionType);
+
+                        // TODO: castExpression.WithAdditionalAnnotations(Simplifier.Annotation) 
+                        // - the Simplifier doesn't remove the redundant cast from the expression
+                        // Issue link: https://github.com/dotnet/roslyn/issues/41500
+                        return currentRoot.ReplaceNode(expression, castExpression.WithAdditionalAnnotations(Simplifier.Annotation));
+                    }
+
+                    return currentRoot;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        private class MyCodeAction : CodeAction.DocumentChangeAction
+        {
+            public MyCodeAction(string title, Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(title, createChangedDocument)
```

Comment:
I believe you need to ensure a non-null equivalence key to ensure correct FixAll.

```suggestion
                : base(title, createChangedDocument, equivalenceKey: title)
```